controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,446 @@
+	}
+	return result, nil
+}
+
+func (r *ManagedFusionDeploymentReconciler) initReconciler(ctx context.Context, req ctrl.Request) {
+	r.ctx = ctx
+
+	r.prometheus = &promv1.Prometheus{}
+	r.prometheus.Name = prometheusName
+	r.prometheus.Namespace = r.Namespace
+
+	r.alertmanager = &promv1.Alertmanager{}
+	r.alertmanager.Name = alertmanagerName
+	r.alertmanager.Namespace = r.Namespace
+
+	r.alertmanagerConfigSecret = &corev1.Secret{}
+	r.alertmanagerConfigSecret.Name = alertmanagerConfigSecretName
+	r.alertmanagerConfigSecret.Namespace = r.Namespace
+
+	r.alertmanagerConfig = &promv1a1.AlertmanagerConfig{}
+	r.alertmanagerConfig.Name = alertmanagerConfigName
+	r.alertmanagerConfig.Namespace = r.Namespace
+
+	r.managedFusionDeploymentSecret = &corev1.Secret{}
+	r.managedFusionDeploymentSecret.Name = managedFusionAgentSecretName
+	r.managedFusionDeploymentSecret.Namespace = r.Namespace
+
+	r.smtpConfigData = &smtpConfig{}
+	r.pagerDutyConfigData = &pagerDutyConfig{}
+}
+
+func (r *ManagedFusionDeploymentReconciler) reconcilePhases() (reconcile.Result, error) {
+	// Uninstallation depends on the status of the components.
+	// We are checking the uninstallation condition before getting the component status
+	// to mitigate scenarios where changes to the component status occurs while the uninstallation logic is running.
+
+	if !r.managedFusionDeploymentSecret.DeletionTimestamp.IsZero() {
+		r.Log.Info("removing finalizer from the managed-fusion-agent-config secret resource")
+		r.managedFusionDeploymentSecret.SetFinalizers(utils.Remove(r.managedFusionDeploymentSecret.GetFinalizers(), managedFusionAgentFinalizer))
+		if err := r.Client.Update(r.ctx, r.managedFusionDeploymentSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from managed-fusion-agent-config secret: %v", err)
+		}
+		r.Log.Info("finallizer removed successfully")
+
+		if err := r.removeOLMComponents(); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to remove agent CSV: %v", err)
+		}
+	} else if r.managedFusionDeploymentSecret.UID != "" {
+		if !utils.Contains(r.managedFusionDeploymentSecret.GetFinalizers(), managedFusionAgentFinalizer) {
+			r.Log.V(-1).Info("finalizer missing on the managed-fusion-agent-config secret resource, adding...")
+			r.managedFusionDeploymentSecret.SetFinalizers(append(r.managedFusionDeploymentSecret.GetFinalizers(), managedFusionAgentFinalizer))
+			if err := r.update(r.managedFusionDeploymentSecret); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update managed-fusion-agent-config secret with finalizer: %v", err)
+			}
+		}
+		if err := r.reconcilePrometheus(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileAlertmanager(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileAlertmanagerConfigSecret(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileAlertmanagerConfig(); err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
Why do we need r.reconcilePrometheusKubeRBACConfigMap, r.reconcilePrometheusService, r.reconcileK8SMetricsServiceMonitor, r.reconcilePrometheusProxyNetworkPolicy functions?

These functions were added for two reasons
1. To prometheus behind TLS
2. Get the metrics from in-cluster monitoring stack

Regarding 1st point, we had prometheus behind TLS to enable ODF dashboard and to let in-cluster monitoring stack federate only the required metrics from storage namespace. I don't think we need any of them in managed-fusion service.
Regarding 2nd point, each offering will have a separate set of metrics that it will need from in-cluster monitoring stack so we should ask offering to implement this functionality in case it is required instead of implementing it for all the offerings.

Reply:
Added all the mentioned functions